pkg/service/mempool.go

```diff
@@ -50,36 +51,58 @@ type flushSpendReq struct {
 
 func NewMempool(storage db.Storage, client eth.Client) *Mempool {
 	return &Mempool{
-		txReqs:     make(chan *txRequest),
-		quit:       make(chan bool),
-		flushReq:   make(chan flushSpendReq),
-		txPool:     make([]MempoolTx, 0),
-		poolSpends: make(map[string]bool),
-		storage:    storage,
-		client:     client,
+		txReqs:           make(chan *txRequest),
+                validatedTxReqs:  make(chan *txRequest),
+		quit:             make(chan bool),
+		flushReq:         make(chan flushSpendReq),
+		txPool:           make([]MempoolTx, 0),
+		poolSpends:       make(map[string]bool),
+		storage:          storage,
+		client:           client,
 	}
 }
 
+func (m *Mempool) validateTransactionRequest(req *txRequest) {
+        tx := req.tx
+        var err error
+        if tx.Body.IsDeposit() {
+                err = m.VerifyDepositTransaction(&tx)
+        } else {
+                // 300-500us
+                err = m.VerifySpendTransaction(&tx)
```

Comment:
Once you move out the `mempool.poolSpends` checks out of this method and it just becomes `validation.ValidateSpendTransaction` (which only checks the previous block in storage and does math only, it basically can be parallelized).

I know you said the `storage.isDoubleSpent` can **not** be parallelized but exactly why? it's checking the previous _persisted_ block only, which won't change until mempool gets flushed. And Mempool basically contains transactions for the next block and having the `ensurePoolSpends` separated on a single thread (under the new `case req := <-m.validatedTxReqs:`) ensures no double-spend happens in mempool.

In summary: `storage` doubleSpend check can be parallelized cause it's past history and static but maintaining mempool double-spend integrity on a single thread keeps the correctness.

Thoughts?

Reply:
Second thoughts on this: realized that if a block is about to be written with transaction X (via flush), but a double-spend transaction X' is in a go-routine just past this line, it's possible for a double-spend to sneak in.

Basically mempool validation can't occur while a block is about to be written.

So you're right - I will need to re-think a little bit more

Reply:
Correct, mempool validation cannot occur while a block is about to be/is being written. That's handled by the `flushReq` case in the mempool loop - all transaction processing halts until the block is successfully written to disk.

Reply:
When I said the double spend check can't be parallelized I was referring to how it requires that blocks be written to disk before new transactions can be processed by the mempool.

Reply:
Yeah but I think there is a race-condition introduced here. As described in my last comment: https://github.com/kyokan/plasma/pull/103/files#r259601343.

The transaction validation can occur simultaneously as a package block write. Basically we should wrap the go-routine in a lock.